### PR TITLE
Dynamic class file deduplication across shims in dist jar build 

### PR DESCRIPTION
**Pull request by gerashegalov**
- uses a binary diff to collect a list of class files identical for all shims in the current build (reduces the jar from 115M to 83M)
- fixes duplicate storage of the shared resources from unshimmed-base, down to 76M 

Numbers are for the biggest jar profile `snapshotsWithDatabricks` 

    
Signed-off-by: Gera Shegalov <gera@apache.org>

### CODE REVIEW

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
build

File: build/buildall [hunk 1]
```diff
@@ -36,16 +36,19 @@ SPARK_SHIM_VERSIONS="
   313
   320
 "
-for sparkver in ${SPARK_SHIM_VERSIONS} ; do
+BUILD_PARALLELISM=1
```
Comment (jlowe) [inline build/buildall:39]: Should this accept an env var override?  e.g.:
```suggestion
BUILD_PARALLELISM=${BUILD_PARALLELISM-1}
```

File: dist/scripts/binary-dedupe.sh [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+# for identical classes.
+
+# We use first shim e.g spark301 as a reference for the following algorithm of identifying
+# identical bytecode across all supported Spark shims:
+# 1. diff spark301 with all other shims pairwise and store the list of identical files
+#    in spark301-spark302.ident
+# 2. count the diff files from Step 1 which is (numShimsInBuild - 1)
+# 3. Call sort on all the diff files and replace duplicate entries with
+#    uniq counts
+# 4. all entries that occur (numShimsInBuild - 1) tinmes are identical for all shims
```
Comment (jlowe) [inline dist/scripts/binary-dedupe.sh:33]:
```suggestion
# 4. all entries that occur (numShimsInBuild - 1) times are identical for all shims
```

File: dist/scripts/binary-dedupe.sh [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+SHIM_DIRS=$(find "$PARALLEL_WORLDS_DIR" -maxdepth 1 -type d -path "*/spark3*" | cut -d/ -f 2)
+REF_SHIM=$(<<< "$SHIM_DIRS" head -1)
+SHIMS_TO_COMPARE=$(<<< "$SHIM_DIRS" tail --lines=+2)
+NUM_DIFFS=$(<<< "$SHIMS_TO_COMPARE" wc -l)
+DIFFDIR=binary-diffs
+
+mkdir $DIFFDIR
+<<< "$SHIMS_TO_COMPARE" xargs -I% -n 1 bash -c \
+  "diff -s -r $PARALLEL_WORLDS_DIR/$REF_SHIM $PARALLEL_WORLDS_DIR/% |
+    grep '.class are identical' |
```
Comment (revans2) [inline dist/scripts/binary-dedupe.sh:49]: nit: In theory we could accidentally put this file into the jar, or have some other file with a line in it that says '.class are identical' which would result in this getting through.  Could we at least make the pattern strict enough that it is highly unlikely anything will get through on accident?

**Review by revans2 — COMMENTED** [review]
Just a small nit.

File: build/buildall (hunk 1 above)
Comment (gerashegalov) [reply build/buildall:39]: at this time we can't allow overriding parallelism. I started the work to make it possible https://github.com/NVIDIA/spark-rapids/pull/3523 showing significant boost. However, there are still issues in the CI left to be resolved.

File: dist/scripts/binary-dedupe.sh (hunk 3 above)
Comment (gerashegalov) [reply dist/scripts/binary-dedupe.sh:49]: agreed, better be defensive

**Review by revans2 — APPROVED** [review]

**Comment by gerashegalov** [comment]
build

**Review by jlowe — APPROVED** [review]

**Pull request merged**